All 14 sprint tasks are completed (-)

maxim-papier opened the pull request.

maxim-papier committed:
Initial commit

maxim-papier committed:
Merge remote-tracking branch 'refs/remotes/origin/main'

maxim-papier committed:
+ Styles, Assets, Launch

maxim-papier committed:
+ Styles, Assets, Launch
+ Tracker cell
+ CollectionView

maxim-papier committed:
+ Header for a section
+ Header label and date label  texts are now truncating correctly
+ TrackerCategory model
+ TrackerRecord model

maxim-papier committed:
+ TabBar

maxim-papier committed:
+ NavBar

maxim-papier committed:
+ Statistic Scene
+ Placeholder object

maxim-papier committed:
+ Placeholder reusable object

maxim-papier committed:
+ Search and filter
+ More tracker mocks

maxim-papier committed:
+ The number of sections and cells in them now depends on the number of categories and trackers in them

maxim-papier committed:
- Close issue "Incorrect behavior during filtering #1"

maxim-papier committed:
+ Filtering by date

maxim-papier committed:
+ Placeholder. It's now shown when search results are empty.
+ If the search query is empty, then filtering is not performed.

maxim-papier committed:
Merge pull request #4 from maxim-papier/sprint_14_datePicker

Sprint 14 date picker

maxim-papier committed:
+ Reusable button object
+ Add tracker or event scene

maxim-papier committed:
Merge pull request #5 from maxim-papier/sprint_14_datePicker

+ Reusable button object

maxim-papier committed:
Stuck. The button does not pressing.

maxim-papier committed:
- fixed button misbehaviours
- changed reusable button object. tapHandler is now necessary when create button, which is pretty logical

maxim-papier committed:
+ Create tracker scene

maxim-papier committed:
+ Colors and emojis are now selected the way they should — only one is active at the moment

maxim-papier committed:
+ Two independent EmojiCell and ColorCell cells. Previously, both sections shared one.
+ Reusing bug fixed
+ Emojis and Color layout bug which occurs when random amount of items comes, fixed

maxim-papier committed:
+ SchedulerVC scene
+ SchedulerCell
+ New logic for listviews. I divide all cells into first, middle, last, and single types. Every type has different styles of rounded corners and separators

maxim-papier committed:
AddScheduler:
- If a toggle is switched on, than the corresponding day of the week will be added to the `selectedDays` array.
- When a user reopens scheduler all toggle will be set in to correct states.

CreateTrackerVC:
- All data from the `selectedDays` array will be copied into the schedule array in the `CreateTrackerVC` when a user dismiss scheduler.

maxim-papier committed:
- The cell color selection has been removed in scheduler
- Some changes in naming

maxim-papier committed:
- Reusable buttons have been completely reengineered.
- `Create New TrackerVC` now saves all data and prepares it for a new tracker creation

maxim-papier committed:
+ The new tracker now saves.
+ The main category array updates safely.
+ etc

maxim-papier committed:
+ doneButton for the trackerCell

maxim-papier committed:
+ add completed trackers to and remove uncompleted from completedTrackers array

maxim-papier committed:
+ Different placeholder shown for search and date picker results
+ Updated placeholder object to enable easy switching between types as needed"
+ Fixed incorrect search results display

maxim-papier committed:
! Fixed issue causing incorrect filter results display after adding a new tracker

IBDesignable commented on TrackerApp/Models/DayOfTheWeek.swift:
```diff
@@ -0,0 +1,23 @@
+
+extension WeekDay {
+    var label: String {
+        return dayLabels[self.rawValue - 1]
```
Смотри что тут мне не нравится:
1) ты завел глобальную константу. Так не принято в swift-мире, поэтому от let dayLabels надо избавляться
2) let dayLabels - ты записал 7 строк и мы надеямся что ты не ошибся в порядке, который здесь используешь. Обычно для енапмов заводят в расширении свойства, как у тебя label, но решают задачу иначе: пишут switch по всем кейсам.

Во первых, это явно задает соответствие и не нужны дополнительные операции взятия по индексу. Во вторых, представим себе какой-нибудь другой enum – например, описание тем в приложении: светлая и темная. Для такого случая можно тоже завести енам и например лейбл, чтобы показывать выбор пользователю. Но если ты забудешь поправить названия в массиве, а количество тем поменяется – добавится новая праздничная тема приложения, то этот способ не будет работать: компилятор не подскажет тебе об ошибке когда ты добавишь новый кейс.

А если перечислять через switch – то любое изменение кейсов будет замечено на уровне компилятора и ты не сможешь сделать ошибку.   Поэтому безопасный путь – переписать здесь на switch

IBDesignable commented on TrackerApp/Models/DayOfTheWeek.swift:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    var shortLabel: String {
+        return shortDayLabels[self.rawValue - 1]
```
аналогично предложение к исправлению

IBDesignable commented on TrackerApp/ObjectsUI/Placeholder.swift:
```diff
@@ -0,0 +1,98 @@
+
+final class Placeholder: UIView {
+
+    var placeholderType: PlaceholderType = .noTrackers {
```
здесь и во всем остальном коде: давай проставим модификатора доступа по намерению использования кода извне

maxim-papier replied:
https://github.com/maxim-papier/TrackerApp/blob/55e425e4a8728b782df4b543b6f80c0fdc62089b/TrackerApp/ObjectsUI/Placeholder.swift#L5-L10
         
Я по изменению значения этой переменной перерисовываю плейсхолдер. Плохая практика?

IBDesignable replied:
нет, вполне ок использовать обсерверы свойств.  

тут главное быть внимательным и случайно не войти в бесконечный цикл вызывов, когда изменение одной переменной, начинает выполнять какой-то код, который снова ее же обновляет

IBDesignable commented on TrackerApp/ObjectsUI/Placeholder.swift:
```diff
@@ -0,0 +1,98 @@
+
+        guard let defaultPlaceholder = UIImage(systemName: "aqi.medium") else { fatalError("SFSymbols image error") }
+
+        let noSearchResultImage = getImage("placeholderNoSearchResults")
+        let noTrackersImage = getImage("placeholderNoTrackers")
+        let noCategoriesImage = getImage("placeholderNoTrackers")
+        let noStatsImage = getImage("placeholderNoStats")
```
давай сделаем два свойства со switch-ом для енама: картинка и текст. 

И тогда твой плейсхолдер сократиться до инита .init(image: risonImage, text: risonTitle)

IBDesignable commented on TrackerApp/Scenes/Cells/ColorCell.swift:
```diff
@@ -0,0 +1,60 @@
+import UIKit
+
+class ColorCell: UICollectionViewCell {
```
Удивлена была реализованным пикерам цвета и эмоджи, ты хорошо поработал 👍🏻

maxim-papier replied:
Сделал случайно, но был рад разобраться с `UICollectionViewCompositionalLayout`. Мне очень понравилось.

IBDesignable commented on TrackerApp/Scenes/SchedulerVC.swift:
```diff
@@ -0,0 +1,137 @@
+
+        cell.labelMenu.text = weekday.label
+
+        switch (indexPath.row, weekdays.count) {
+
+        case (firstIndex, 1): cell.buttonPosition = .single
+        case (firstIndex, _): cell.buttonPosition = .first
+        case (lastIndex, _): cell.buttonPosition = .last
+        default: cell.buttonPosition = .middle
+        }
```
Здесь хорошо бы разбивать код на функции чтобы улучшить понимания что здесь происходит. Лучше разбивать алгоритм на чистые функции.

Чистая функция - это функция, которая возвращает результат на основе ее входных аргументов и не имеет побочных эффектов, то есть она не изменяет состояние программы и не взаимодействует с внешними ресурсами. Параметры на вход – и ожидаемый результат на выходе.

maxim-papier replied:
Создал issue

IBDesignable commented on TrackerApp/Scenes/TrackerVС.swift:
```diff
@@ -0,0 +1,340 @@
+        cell.doneButton.backgroundColor = show.color
+        cell.titleLabel.text = show.title
+        cell.emojiLabel.text = show.emoji
+        cell.daysLabel.text = "XX дней"
```
Лучше поставить 0, можно не реализовывать логику инкремента сейчас, но так будем попадать в макеты

IBDesignable commented on TrackerApp/Scenes/TrackerVС.swift:
```diff
@@ -0,0 +1,340 @@
+        guard let textInput = searchController.searchBar.text else { return }
+        searchText = textInput
+
+        switch searchText.isEmpty {
+        case true: break
+        case false: filterResults(with: searchText)
+        }
```
во первых, для true/false прекрасно подходит оператор if-else.

Во вторых, раз один из кейсов не нужен, было бы локаничнее написать if !searchText.isEmpty { .... }

IBDesignable commented on TrackerApp/Scenes/Cells/EmojiCell.swift:
```diff
@@ -0,0 +1,46 @@
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        super.layoutSubviews()
```
1) давай не будем вызывать метод super не из переопределяемого метода.  переопределяешь инит – ок, super init обязателен. переопределяешь layoutSubviews – аналогично. Иначе ты преждевременно вызываешь метод жизненного цикла вью, который возможно будет аффектить твое приложение непредсказуемым образом. 
2) если ты хочешь попросить вью отрендериться верно: то лучше вызывай layoutIfNeed() - если вью будет нужна, она сама перестроится

Но на самом деле, здесь этот код не нужен и строчку эту надо удалить. В приложении есть еще несколько таких ошибок: давай исправим

IBDesignable commented on TrackerApp/Extensions/ShowPreviewUI.swift:
```diff
@@ -0,0 +1,15 @@
+
+extension UIViewController {
+
+    private struct Preview: UIViewControllerRepresentable {
+        let vc: UIViewController
+
+        func makeUIViewController(context: Context) -> some UIViewController { vc }
+        func updateUIViewController(_ uiViewController: UIViewControllerType, context: Context) { }
+    }
+
+    func showPreview() -> some View {
+        Preview(vc: self).edgesIgnoringSafeArea(.all)
+    }
```
Очень прикольно! мне понравился такой хак :) буду благодарна за референс)

maxim-papier replied:
Да, вещь - https://swiftjectivec.com/Using-Xcode-Previews-for-UIKit/

IBDesignable replied:
Спасибо!

IBDesignable commented on TrackerApp/Extensions/ColorsYP.swift:
```diff
@@ -0,0 +1,42 @@
+}
+
+enum SelectionColorStyle: String, CaseIterable {
+    case selection01 = "selection-01"
+    case selection02 = "selection-02"
+    case selection03 = "selection-03"
+    case selection04 = "selection-04"
+    case selection05 = "selection-05"
+    case selection06 = "selection-06"
+    case selection07 = "selection-07"
+    case selection08 = "selection-08"
+    case selection09 = "selection-09"
+    case selection10 = "selection-10"
+    case selection11 = "selection-11"
+    case selection12 = "selection-12"
+    case selection13 = "selection-13"
+    case selection14 = "selection-14"
+    case selection15 = "selection-15"
+    case selection16 = "selection-16"
+    case selection17 = "selection-17"
+    case selection18 = "selection-18"
```
вообще не понятно что это такое по коду. Уточнить бы название SelectionColorStyle, и возможно задать кейсам имена симантичнее - например чтобы было понятно где здесь красный. 

Это могло бы пригодится, если ты захочешь добавить поддержку VoiceOver в приложении: тогда придется писать понятные кейсы.
[Здесь можно книжку ](https://rubanov.dev/a11y-book/)посмотреть Михаила Рубанова – очень известный программист в этой области, даже среди зарубежного сообщества

maxim-papier replied:
Я такие имена дал, потому, что они совпадают с теми, которые дал дизайнер в Figma. Как принято в такой ситуации поступать, следовать за неймингом, который определен в макете или делать более понятные свои названия? 

Спасибо за книжку. По оглавлению очень привлекательно. Буду читать.

IBDesignable replied:
У дизайнеров обычно такая же проблема с неймингом, как и у разработчиков, поэтому тут обычно если что-то не нравится – стараются договориться и найти общее решение. И действительно, удобнее иметь консистентные с дизайном названия – ты тут прав, я полностью согласна что это хорошая стратегия.

Если хочешь этого правила придерживаться здесь – окей, все равно нет возможности пообщаться с дизайнером (так и в жизни бывает). 

Но с точки зрения читаемости кода, получилось "не удобно". Через месяц ты и не вспомнишь почему такое решение принято и почему все так пронумеровано. Ведь можно было бы вместо енама использовать что-то вроде 0...20.map { "selection-\($0)" }. И получилось бы короче, и задачу бы решало. 

А если тебе нужна функциональность enum - то лучше все-таки придерживаться читаемых/говорящих названий – будущему себе скажешь спасибо за это :)

maxim-papier replied:
Большое спасибо за информацию. Книжку начал читать. Меня, вообще, доступность интересует из-за дизайнерского бэкграунда.

maxim-papier committed:
Removed unnecessary super.layoutSubviews()
Fixed a bug in displaying text on buttons in dark mode
Optimized placeholder object
Removed global constants from Weekday object
